File: day04.py
#!/usr/bin/env pipenv run python
"""Solutions to day 4 of Advent of Code"""

import re
import itertools
import collections

from get_input import get_input, line_parser

def part1(guard_log):
    """Solution to part 1"""
    guard_dict = collections.defaultdict(list)
    current_guard = None
    is_asleep = False
    for year, month, day, hour, minute, guard, message in sorted(guard_log, key=lambda g: g[:5]):
        if guard is not None:
            assert not is_asleep
            current_guard = guard
            is_asleep = False
        elif message == "falls asleep":
            assert not is_asleep
            is_asleep = True
            start = (year, month, day, hour, minute)
        elif message == "wakes up":
            assert is_asleep
            is_asleep = False
            end = (year, month, day, hour, minute)
            assert start[:3] == end[:3]
            time = 60*(end[3] - start[3])+end[4]-start[4]
            guard_dict[current_guard].append((time, tuple(start), tuple(end)))
        else:
            raise Exception(f"Unrecognized message {message}")
    most_rested, times = max(guard_dict.items(), key=lambda kv: sum(t[0] for t in kv[1]))
    
    time_counter = collections.defaultdict(int)
    for hour, minute in itertools.product(range(24), range(60)):
        for _, start, end in times:
            if start[3] <= hour < end[3] or (start[3] == end[3] and start[4] <= minute < end[4]):
                time_counter[(hour, minute)] += 1
    result = max(time_counter.items(), key=lambda kv: kv[1])
    return result[0][1] * most_rested

def part2(guard_log):
    """Solution to part 2"""
    guard_dict = collections.defaultdict(list)
    current_guard = None
    is_asleep = False
    for year, month, day, hour, minute, guard, message in sorted(guard_log, key=lambda g: g[:5]):
        if guard is not None:
            assert not is_asleep
            current_guard = guard
            is_asleep = False
        elif message == "falls asleep":
            assert not is_asleep
            is_asleep = True
            start = (year, month, day, hour, minute)
        elif message == "wakes up":
            assert is_asleep
            is_asleep = False
            end = (year, month, day, hour, minute)
            assert start[:3] == end[:3]
            time = 60*(end[3] - start[3])+end[4]-start[4]
            guard_dict[current_guard].append((time, tuple(start), tuple(end)))
        else:
            raise Exception(f"Unrecognized message {message}")
    time_counter = collections.defaultdict(int)
    for hour, minute in itertools.product(range(24), range(60)):
        for guard, times in guard_dict.items():
            for _, start, end in times:
                if start[3] <= hour < end[3] or (start[3] == end[3] and start[4] <= minute < end[4]) and not (hour == end[3] and minute == end[4]):
                    time_counter[(guard, hour, minute)] += 1
    guard, _ = max(time_counter.items(), key=lambda kv: kv[1])
    return guard[0] * guard[2]


def parse(line):
    # format 
    # [1518-11-01 00:00] Guard #10 begins shift
    match = re.match(r'\[(\d+)-(\d+)-(\d+) (\d+):(\d+)](?: Guard #(\d+))? (.*)$', line)
    return [int(g) if (g is not None and g.isdigit()) else g for g in match.groups()]

if __name__ == '__main__':
    LINES = line_parser(get_input(day=4, year=2018), parse=parse)
    print("Part 1: {}".format(part1(LINES)))
    print("Part 2: {}".format(part2(LINES)))

File: day04.rs
extern crate common;
extern crate regex;

use std::ops::Sub;
use std::cmp::Ordering;
use std::collections::HashMap;

use common::get_input;
use regex::Regex;

type GuardId = u16;

#[derive(Clone, Debug)]
enum EventType {
    WakesUp,
    FallsAsleep,
    GuardChange(GuardId),
}

impl EventType {
    fn new(event: &str) -> EventType {
        let guard = Regex::new(r"Guard #(\d+) begins shift").unwrap();
        match event {
            "falls asleep" => EventType::FallsAsleep,
            "wakes up" => EventType::WakesUp,
            _ => {
                let g = guard.captures(event).unwrap();
                EventType::GuardChange(g[1].parse().unwrap())
            }
        }
    }
}

#[derive(Clone, Debug, PartialEq, Eq)]
struct Date {
    year: u16,
    month: u8,
    day: u8,
    hour: u8,
    minute: u8,
}

impl Date {
    fn new(year: &str, month: &str, day: &str, hour: &str, minute: &str) -> Date {
        return Date {
            year: year.parse().unwrap(),
            month: month.parse().unwrap(),
            day: day.parse().unwrap(),
            hour: hour.parse().unwrap(),
            minute: minute.parse().unwrap(),
        }
    }

    fn hour_diff(&self, other: &Date) -> u32 {
        return (u32::from(other.hour) - u32::from(self.hour)) * 60 +
            u32::from(other.minute) - u32::from(self.minute);
    }
}

#[derive(Clone, Debug)]
struct Event {
    date: Date,
    event_type: EventType,
}

impl PartialOrd for Date {
    fn partial_cmp(&self, other: &Date) -> Option<Ordering> {
        return Some(self.cmp(&other));
    }
}

impl Ord for Date {
    fn cmp(&self, other: &Date) -> Ordering {
        if self.year != other.year { return self.year.cmp(&other.year) }
        if self.month != other.month { return self.month.cmp(&other.month) }
        if self.day != other.day { return self.day.cmp(&other.day) }
        if self.hour != other.hour { return self.hour.cmp(&other.hour) }
        if self.minute != other.minute { return self.minute.cmp(&other.minute) }
        return Ordering::Equal
    }
}

fn part1(guard_events: &HashMap<GuardId, Vec<(Date, Date)>>) -> u32 {
    let (guard, times) = guard_events
        .iter()
        .max_by(|(a, b)| {
            let mut total = 0;)
            for (start, stop) in times {
                total += start.hour_diff(&stop);
            }
            return total;
        })
        .unwrap();
    unimplemented!();
}

fn part2(events: &HashMap<GuardId, Vec<(Date, Date)>>) -> String {
    unimplemented!()
}

fn parse<'a>(lines: String) -> HashMap<GuardId, Vec<(Date, Date)>> {
    let event = Regex::new(r"\[(\d{4})\-(\d{2})\-(\d{2}) (\d{2}):(\d{2})\] (.*)$").unwrap();
    let mut events: Vec<Event> = Vec::new();
    for line in lines.trim().split('\n') {
        let m = event.captures(line).unwrap();
        events.push( Event {
            date: Date::new(&m[1], &m[2], &m[3], &m[4], &m[5]),
            event_type: EventType::new(&m[6]),
        })
    }
    events.sort_unstable_by_key(|a| a.date.clone());
    let mut result: HashMap<GuardId, Vec<(Date, Date)>> = HashMap::new();
    let mut guard: Option<GuardId> = None;
    let mut sleep_start: Option<Date> = None;
    for event in events {
        match event.event_type {
            EventType::GuardChange(id) => guard = Some(id),
            EventType::FallsAsleep => sleep_start = Some(event.date),
            EventType::WakesUp => {
                result.entry(guard.unwrap())
                   .or_insert(Vec::new())
                   .push((sleep_start.unwrap(), event.date));
                sleep_start = None;
            }
        }
    }
    return result;
}

fn main() {
    let input = parse(get_input(04, 2018));
    println!("Part 1: {}", part1(&input));
    println!("Part 2: {}", part2(&input));
}
